OpenKeychain/src/main/java/org/sufficientlysecure/keychain/remote/SSHAgentService.java

```diff
@@ -0,0 +1,341 @@
+import android.content.Intent;
+import android.os.IBinder;
+import android.util.Log;
+
+import org.bouncycastle.bcpg.HashAlgorithmTags;
+import org.openintents.ssh.ISSHAgentService;
+import org.openintents.ssh.SSHAgentApiConstants;
+import org.sufficientlysecure.keychain.Constants;
+import org.sufficientlysecure.keychain.pgp.CanonicalizedPublicKey;
+import org.sufficientlysecure.keychain.pgp.CanonicalizedPublicKeyRing;
+import org.sufficientlysecure.keychain.pgp.exception.PgpGeneralException;
+import org.sufficientlysecure.keychain.ssh.AuthenticationData;
+import org.sufficientlysecure.keychain.ssh.AuthenticationOperation;
+import org.sufficientlysecure.keychain.operations.results.OperationResult.LogEntryParcel;
+import org.sufficientlysecure.keychain.ssh.AuthenticationResult;
+import org.sufficientlysecure.keychain.pgp.exception.PgpKeyNotFoundException;
+import org.sufficientlysecure.keychain.provider.ApiDataAccessObject;
+import org.sufficientlysecure.keychain.provider.CachedPublicKeyRing;
+import org.sufficientlysecure.keychain.provider.KeyRepository;
+import org.sufficientlysecure.keychain.provider.KeychainContract;
+import org.sufficientlysecure.keychain.service.input.CryptoInputParcel;
+import org.sufficientlysecure.keychain.service.input.RequiredInputParcel;
+import org.sufficientlysecure.keychain.ssh.AuthenticationParcel;
+
+import java.security.PublicKey;
+import java.util.Date;
+import java.util.HashSet;
+
+
+public class SSHAgentService extends Service {
+    public static final String TAG = "SSHAgentService";
+
+    private ApiPermissionHelper mApiPermissionHelper;
+    private KeyRepository mKeyRepository;
+    private ApiDataAccessObject mApiDao;
+    private ApiPendingIntentFactory mApiPendingIntentFactory;
+
+    @Override
+    public void onCreate() {
+        super.onCreate();
+        mApiPermissionHelper = new ApiPermissionHelper(this, new ApiDataAccessObject(this));
+        mKeyRepository = KeyRepository.create(this);
+        mApiDao = new ApiDataAccessObject(this);
+
+        mApiPendingIntentFactory = new ApiPendingIntentFactory(getBaseContext());
+    }
+
+    protected final ISSHAgentService.Stub mSSHAgent = new ISSHAgentService.Stub() {
+        @Override
+        public Intent execute(Intent intent) {
+            Log.d(TAG, "===>>> execute STUB: " + intent.getAction());
+            return checkIntent(intent);
+        }
+
+    };
+
+    private Intent checkIntent(Intent intent) {
+        Intent errorResult = checkRequirements(intent);
+        if (errorResult == null) {
+            return executeIntentInternal(intent);
+        } else {
+            return errorResult;
+        }
+    }
+
+    private Intent executeIntentInternal(Intent intent) {
+        Log.d(TAG, "===>>> executeIntentInternal 0: " + intent);
+        switch (intent.getAction()) {
+            case SSHAgentApiConstants.ACTION_SIGN:
+                return sign(intent);
+            case SSHAgentApiConstants.ACTION_SELECT_KEY:
```

Comment:
same for the API: I would propose we simply call this ``SshApi``

Reply:
I agree, I'm working on this right now and will move the constants into a new class SSHApi and add some helper functions for versioning and async execution.
(Altough, SSHApi sounds a bit generic, but it seems the best tradeoff)

Reply:
Acronyms aren't capitalized, please make that SshApi

Reply:
As an explanation: Bouncy Castle has the coding style to use upper case acronyms. But in Android and in OpenKeychain acronyms are not capitalized. So that's why in some parts of the code it is mixed style.